lib/package_storage.rb

```diff
@@ -0,0 +1,52 @@
+# frozen_string_literal: true
+
+# PackageStorage is a factory for proxies to a Package's underlying storage.
+# An instance supports a named set of format types, mapping each to a storage
+# proxy class and creating an appropriate proxy instance for a package.
+class PackageStorage
+
+  # Create a PackageStorage instance.
+  #
+  # TODO: unify VolumeManager and PackageStorage -- Volume should bind not only
+  # its format name, but the storage proxy class; then this should take the volumes
+  # and the volume manager is no longer needed. We only use volumes to encapsulate
+  # the actual storage and storage proxy classes, so they should not leak from here.
+  #
+  # @param formats [Hash<Symbol, Class>] a mapping of the format name to proxy class
+  # @param volume_manager [VolumeManager] the VolumeManager for the Volumes where
+  #   packages are stored
+  # @example
+  #   PackageStorage.new(formats: { bag: Chipmunk::Bag }, volume_manager: a_volume_manager)
+  def initialize(formats:, volume_manager:)
+    @formats = formats
+    @volume_manager = volume_manager
+  end
+
+  # Retrieve the storage proxy for a given package. The package must be stored and of a registered format.
+  # @param package [Package] the package for which to retrieve a storage proxy
+  # @raise [Chipmunk::PackageNotStoredError] if the package is not stored
+  # @raise [Chipmunk::UnsupportedFormatError] if the package format is not registered
+  def for(package)
+    raise Chipmunk::PackageNotStoredError, package unless package.stored?
+
+    storage_for(package.format).new(path_to(package))
```

Comment:
This is kind of the key to the whole refactoring; trying to unpack this...
`storage_for(package.format)` for the bag case would return a `Chipmunk::Bag`
the volume's responsibility is currently only to map from `package.storage_path` to an absolute path

so we have decoupled the storage format (i.e. storage proxy class) from the volume

but I think the TODO is about re-coupling that: "Volume should bind not only its format name, but the storage proxy class (i.e. Chipmunk::Bag)"

Reply:
As above, we agree that it's a little odd to call this thing "storage for", but "proxy class for", etc., are not much better. This probably improves when the Volume is bound to its format, but we still have to name what this is... It may be Format (and a format may require Storage), or it may really be a Package as we pull up the notion of an Artifact separate from its stored representation. Too early to say.

Reply:
Binding the Volume to the storage format allowed this to read considerably better:

`volume_for(package).get(package.storage_path)`